sdk/src/lib.rs

```diff
@@ -0,0 +1,328 @@
+        pub mod v1alpha2 {
+            tonic::include_proto!("apibara.node.v1alpha2");
+
+            pub(crate) const FILE_DESCRIPTOR_SET: &[u8] =
+                tonic::include_file_descriptor_set!("starknet_descriptor_v1alpha2");
+
+            pub fn stream_file_descriptor_set() -> &'static [u8] {
+                FILE_DESCRIPTOR_SET
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Configuration {
+    batch_size: u64,
+    starting_cursor: Option<Cursor>,
+    finality: Option<i32>,
+    filter: Option<Filter>,
+}
+impl Configuration {
+    pub fn new(
+        batch_size: u64,
+        starting_cursor: Option<Cursor>,
+        finality: Option<i32>,
+        filter: Option<Filter>,
+    ) -> Self {
+        Self {
+            batch_size,
+            starting_cursor,
+            finality,
+            filter,
+        }
+    }
+}
+impl Default for Configuration {
+    fn default() -> Self {
+        Self {
+            batch_size: 1,
+            starting_cursor: None,
+            finality: None,
+            filter: Some(Filter::default()),
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ApibaraIndexerError {
+    #[error("Failed to build indexer")]
+    FailedToBuildIndexer,
+    #[error(transparent)]
+    TonicError(#[from] tonic::transport::Error),
+    #[error(transparent)]
+    FailedToConfigureStream(#[from] SendError<Configuration>),
+    #[error(transparent)]
+    StreamError(#[from] tonic::Status),
+}
+
+pub struct ApibaraIndexer<StreamClient> {
+    client: Option<StreamClient>,
+    configuration: Option<Configuration>,
+    configuration_channel: Option<(Sender<Configuration>, Receiver<Configuration>)>,
+    inner_channel: Option<(Sender<StreamDataRequest>, Receiver<StreamDataRequest>)>,
+}
+
+impl<StreamClient> Default for ApibaraIndexer<StreamClient> {
+    fn default() -> Self {
+        Self {
+            client: None,
+            configuration: None,
+            configuration_channel: None,
+            inner_channel: None,
+        }
+    }
+}
+
+pub struct StreamHandler {
+    stream_id: u64,
+    configuration_rx: Receiver<Configuration>,
+    inner: StreamingClientWrapper,
+    inner_tx: Sender<StreamDataRequest>,
+}
+
+#[derive(Debug)]
+#[pin_project]
+struct StreamingClientWrapper {
+    #[pin]
+    inner: Streaming<StreamDataResponse>,
+}
+
+impl ApibaraIndexer<StreamClient<Channel>> {
+    pub fn with_client(mut self, client: StreamClient<Channel>) -> Self {
+        self.client = Some(client);
+        self
+    }
+
+    pub fn with_configuration(mut self, configuration: Configuration) -> Self {
+        self.configuration = Some(configuration);
+        self
+    }
+    pub fn with_configuration_channel(
+        mut self,
+        configuration_channel: (Sender<Configuration>, Receiver<Configuration>),
+    ) -> Self {
+        self.configuration_channel = Some(configuration_channel);
+        self
+    }
+
+    pub fn with_inner_channel(
```

Comment:
This is an implementation detail, so no need to add a builder for it. Remove the fn.